src/index.js

```diff
@@ -0,0 +1,250 @@
+}
+
+function runRefreshTokenPromise() {
+  refreshAccessTokenPromise = new Promise((resolve, reject) => {
+    // fetch new token with refresh token
+    return Promise.resolve(config.prepareRefreshTokenRequest(tokens.refreshToken))
+      .then(tokenRequest => fetch(tokenRequest))
+      .then(response => {
+        refreshAccessTokenPromise = null;
+
+        if (response.status !== STATUS_OK) {
+          throw new Error(ERROR_REFRESH_TOKEN_EXPIRED);
+        }
+
+        return config.getAccessTokenFromResponse(response);
+      })
+      // save access token to local config
+      .then(token => {
+        tokens.accessToken = token;
+
+        if (config.onAccessTokenAcquired) {
+          config.onAccessTokenAcquired(token);
+        }
+
+        resolve(token);
+      })
+      .catch(error => {
+        refreshAccessTokenPromise = null;
+
+        tokens.accessToken = null;
+        tokens.refreshToken = null;
+
+        reject(error);
+      });
+  });
+  return refreshAccessTokenPromise;
+}
+
+function convertToRequest(args) {
+  const request = new Request(...args);
+  request.id = Date.now();
+  return request;
+}
+
+function shouldFetchAccessToken(request) {
+  // check if we're already fetching the token
+  if (refreshAccessTokenPromise) {
+    return false;
+  }
+
+  const requestAccessToken = parseBearer(request.headers.get('authorization'));
+  if (requestAccessToken !== tokens.accessToken) {
+    return false;
+  }
+
+  return true;
+}
+
+function isAuthorized() {
+  return !!tokens.refreshToken;
+}
+
+function fetchInterceptor(fetch, ...args) {
+  if (!config.shouldIntercept ||
+    !config.setRequestAuthorization ||
+    !config.prepareRefreshTokenRequest ||
+    !config.getAccessTokenFromResponse) {
+    throw new Error(ERROR_INVALID_CONFIG);
+  }
+
+  const request = convertToRequest(args);
+
+  // check whether we should ignore this request
+  if (!isAuthorized() || !config.shouldIntercept(request)) {
+    return fetch(request);
+  }
+
+  // outer fetch promise
+  return new Promise((outerResolve, outerReject) => {
+    // inner promise which includes resolving access token
+    const runInnerPromise = () =>
+      Promise.resolve(request)
+        .then((request) => config.setRequestAuthorization(request, tokens.accessToken))
+        // initial fetch
+        .then(request => fetch(request))
+        .then(response => {
+          // check if response invalidates access token
+          if (config.shouldInvalidateAccessToken) {
+            return Promise.all([response, config.shouldInvalidateAccessToken(response)]);
+          }
+
+          return response;
+        })
+        .then(results => {
+          const response = results[0];
+          const shouldInvalidateAccessToken = results[1];
+
+          if (shouldInvalidateAccessToken) {
+            tokens.accessToken = null;
+            runRefreshTokenPromise();
+          }
+
+          return response;
+        })
+        .then(response => {
```

Comment:
Please check usage of `.then` sometimes it's not necessary, I think once you encapsulate logic into methods that need for `.then` will diminish. You explained me concept of config methods that returns promise, and for that cases you don't have other options.